tiledb/api/src/column.rs

```diff
@@ -0,0 +1,509 @@
+                    ))
+                },
+                None => None,
+            },
+        }
+    }
+
+    pub fn is_allocated(&self) -> bool {
+        match self {
+            Column::Referenced(_) => false,
+            Column::Allocated(_) => true,
+        }
+    }
+
+    pub fn to_json(
+        &self,
+        datatype: Datatype,
+        cell_val_num: u32,
+    ) -> TileDBResult<JsonValue> {
+        let is_strings =
+            matches!(datatype, Datatype::StringAscii | Datatype::StringUtf8);
+
+        return if cell_val_num == u32::MAX && is_strings {
+            let result = self.to_string_vec(datatype, cell_val_num);
+            if result.is_ok() {
+                Ok(json!(result.unwrap()))
+            } else {
+                Ok(json!(self.to_boxed_u8_vec(datatype, cell_val_num)?))
+            }
+        } else if cell_val_num == u32::MAX {
+            Ok(json!(self.to_boxed_u8_vec(datatype, cell_val_num)?))
+        } else {
+            Ok(fn_typed!(self.to_vec, datatype, datatype, cell_val_num =>
+                json!(to_vec?)
+            ))
+        };
+    }
+}
+
+impl<'c1, 'c2> PartialEq<Column<'c2>> for Column<'c1> {
+    fn eq(&self, other: &Column<'c2>) -> bool {
+        if self.num_values() != other.num_values() {
+            return false;
+        }
+
+        if self.value_size() != other.value_size() {
+            return false;
+        }
+
+        if self.data() != other.data() {
+            return false;
+        }
+
+        let offsets_match = match (self.offsets(), other.offsets()) {
+            (Some(mine), Some(theirs)) => mine == theirs,
+            _ => false,
+        };
+        if !offsets_match {
+            return false;
+        }
+
+        true
+    }
+}
+
+/// The AsColumn trait provides an auto-conversion from common datatypes into
+/// the columnar format used by TileDB.
+///
+/// There are three type "patterns" that fit into the auto conversion scheme.
+///
+/// 1. The simplest is a Vec or slice of primitive values. This mode can be
+///    used for any non-variable sized cell type, even multi-value cells. The
+///    length of the Vec or slice must be a multiple of the cell value num.
+///
+/// 2. The second most common is for string based data. This allows users to
+///    pass a Vec or slice of `AsRef<str>` instances which are then transformed
+///    into the correct data and offsets buffers. This method allocates memory
+///    to hold the linearized data buffer and offsets buffer.
+///
+/// 3. The third is a pair of Vec's or slices where the first is the contents
+///    of the data buffer and the second is a `Vec<u64>` or `&[u64]` that
+///    contains the offsets for each var sized cell. Generally speaking this
+///    is an uncommon use case when users need a variable sized cell that is
+///    not a string.
+pub trait AsColumn<'data> {
+    fn as_column(&'data self) -> Column<'data>;
+}
+
+pub trait TryAsColumn<'data> {
```

Comment:
This is a more specialized `TryInto` ([doc](https://doc.rust-lang.org/std/convert/trait.TryInto.html)), probably better just to do `impl<'data, T: AsColumn<'data>> TryInto<Column<'data>> for T`